pkg/kubelet/pod_workers.go

```diff
@@ -913,15 +914,20 @@ func (p *podWorkers) managePodLoop(podUpdates <-chan podWork) {
 			case update.Options.RunningPod != nil:
 				// when we receive a running pod, we don't need status at all
 			default:
-				// wait until we see the next refresh from the PLEG via the cache (max 2s)
-				// TODO: this adds ~1s of latency on all transitions from sync to terminating
-				//  to terminated, and on all termination retries (including evictions). We should
-				//  improve latency by making the the pleg continuous and by allowing pod status
-				//  changes to be refreshed when key events happen (killPod, sync->terminating).
-				//  Improving this latency also reduces the possibility that a terminated
-				//  container's status is garbage collected before we have a chance to update the
-				//  API server (thus losing the exit code).
-				status, err = p.podCache.GetNewerThan(pod.UID, lastSyncTime)
+				if !pleg.IsEventedPLEGInUse {
```

Comment:
This is not correct, the point of the podCache GetNewerThan is to enforce consistency AND to ensure that we don't double process events.  If relist is still being called, it is not safe to directly invoke the cache.

Roughly, GetNewerThan is still correct even assuming an evented PLEG, and instead we need to ensure that the podCache is consistent.

Reply:
Oh ok, need to take a closer at this. In my testing if we use `GetNewerThan` with evented pleg it would lead to missing states. e.g containers would go directly from creating to completed without ever going through `Running`. I will spend some more time on this `GetNewerThan` interface to understand it better.

Reply:
It probably means you aren’t generating all the same events in the same order (for any given container).  While the total order won’t be the same, the sequence for each pod should be.

I would recommend instrumenting both PLEG implementations and having them output a list of all of the events they both generate, just to be sure.  This is what I meant by “improve consistency checking” - just debugging the issues you hit (like GetNewerThan) will become easier because you can rule out the PLEG as the cause.